lpc176x: read chip id and use as usb serial.

mattthebaker opened the pull request:
Signed-off-by: Matt Baker <baker.matt.j@gmail.com>

mattthebaker commented:
Looking for some feedback on the implementation for cleanup.

One improvement might be more explicit passing of length between the serial functions. Also having Kconfig disable the user serial number when the chipid serial number is enabled.

KevinOConnor commented on src/lpc176x/usbserial.c:
```diff
@@ -16,6 +16,7 @@
 #include "sched.h" // DECL_INIT
 #include "usb_cdc_ep.h" // USB_CDC_EP_BULK_IN
+#include "uid.h" // platform_get_uid
 
```
I'd say move `usb_set_serial()`, uid.h, and uid.c all into src/lpc176x/usbserial.c.

mattthebaker replied:
Sounds good and done. When I start adding other platforms usb_set_serial, or at least the part of the function that breaks a byte array into hex string may need to be accessible by other platforms' code. We can save that for future pull request.

mattthebaker commented on src/Kconfig:
```diff
@@ -63,6 +67,10 @@ config USB_DEVICE_ID
 config USB_SERIAL_NUMBER
     string "USB serial number"
```
Done

mattthebaker commented on src/Kconfig:
```diff
@@ -63,6 +67,10 @@ config USB_DEVICE_ID
+config USB_SERIAL_NUMBER_CHIPID
+    depends on CHIPID_LEN != 0
+    bool "USB serial number from CHIPID"
+    default 0
```
The length is now hidden in usbserial.c for the platform so I changed this to boolean.

mattthebaker commented on src/generic/usb_cdc.c:
```diff
@@ -365,11 +376,18 @@ usb_req_get_descriptor(struct usb_ctrlrequest *req)
+                && READP(d->wValue) == ((USB_DT_STRING<<8) | USB_STR_ID_SERIAL)
+                && READP(d->wIndex) == USB_LANGID_ENGLISH_US) {
+                    size = SIZE_cdc_string_serial_chipid;
+                    desc = &cdc_string_serial_chipid;
```
Sounds good and done. I felt usb_req_get_descriptor needed a tiny refactor for it, you'll see it in the code.

KevinOConnor commented on src/generic/usb_cdc.c:
```diff
@@ -358,21 +358,34 @@ usb_req_get_descriptor(struct usb_ctrlrequest *req)
+        && req->wIndex == USB_LANGID_ENGLISH_US) {
+            struct usb_string_descriptor *usbserial_serialid;
+            usbserial_serialid = usbserial_get_serialid();
+            size = usbserial_serialid->bLength;
```
I know this is directly contradicting my previous comments, but it occurred to me another way to do this would be to introduce two new platform specific calls: `uint_fast8_t usbserial_serialid_length(void)` and `void usbserial_fill_serialid(uint16_t *data)`.  Then the generic code could do:
```
    size = get_data_length();
    struct {
        struct usb_string_descriptor desc;
        uint16_t data[size];
    } ldesc;
    ldesc.desc.bLength = sizeof(ldesc.desc) + 2 * size;
    ldesc.desc.bDescriptorType = USB_DT_STRING;
    usbserial_fill_serialid(ldesc.data);
```

The main advantage of this approach is that it is not necessary to allocate global memory for the serial id.  Which may be useful on some platforms starved for ram.

I can make this change though, as I don't want you to have to go through multiple revisions.

mattthebaker commented:
Cleaned up per previous comments. I had refactored usb_set_serial for the stack version already, so I also added what i think is slightly cleaner code for that part.

KevinOConnor commented:
Thanks!

FYI, I also went on to change the "product id" from "Klipper firmware" to the mcu name.

-Kevin

mattthebaker commented:
That sounds great, thanks!

Matt
